fuse_core/include/fuse_core/util.hpp

```diff
@@ -150,6 +151,178 @@ Eigen::Matrix<T, 2, 2, Eigen::RowMajor> rotationMatrix2D(const T angle)
+        1.0) *
+        (2.0 * qy * qy + 2.0 * qz * qz - 1.0));
+      jacobian_map(2, 1) =
+        -(2.0 * qy) /
+        ((std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) +
+        1.0) *
+        (2.0 * qy * qy + 2.0 * qz * qz - 1.0));
+      jacobian_map(2, 2) =
+        -((2.0 * qx) / (2.0 * qy * qy + 2.0 * qz * qz - 1.0) -
+        (4.0 * qy * (2.0 * qw * qz + 2.0 * qx * qy)) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) + 1.0);
+      jacobian_map(2, 3) =
+        -((2.0 * qw) / (2.0 * qy * qy + 2.0 * qz * qz - 1.0) -
+        (4.0 * qz * (2.0 * qw * qz + 2.0 * qx * qy)) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0)) /
+        (std::pow((2.0 * qw * qz + 2.0 * qx * qy), 2.0) / std::pow((2.0 * qy * qy + 2.0 * qz * qz - 1.0), 2.0) + 1.0);
+    }
+  }
+}
+
+/**
+ * @brief Compute product of two quaternions and the function jacobian
+ * TODO(giafranchini): parametric jacobian computation? Atm this function is only used in 
+ * normal_prior_pose_3d cost function. There we only need the derivatives wrt quaternion W,
+ * so at the time we are only computing the jacobian wrt W
+ * 
+ * @param[in] z Pointer to the first quaternion array (4x1)
+ * @param[in] w Pointer to the second quaternion array (4x1)
+ * @param[in] z Pointer to the first quaternion array (4x1) 
+ * @param[in] jacobian Pointer to the jacobian matrix (4x4, optional)
+ */
+static inline void quaternionProduct(const double * z, const double * w, double * zw, double * jacobian = nullptr) 
+{
+  ceres::QuaternionProduct(z, w, zw);
+  if (jacobian) {
+    Eigen::Map<Eigen::Matrix<double, 4, 4, Eigen::RowMajor>> jacobian_map(jacobian);
+    jacobian_map << 
+      z[0], -z[1], -z[2], -z[3],
+      z[1],  z[0], -z[3],  z[2],
+      z[2],  z[3],  z[0], -z[1],
+      z[3], -z[2],  z[1],  z[0];
+  }
+}
+
+/**
+ * @brief Compute quaternion to AngleAxis conversion and the function jacobian
+ * 
+ * @param[in] q Pointer to the quaternion array (4x1)
+ * @param[in] angle_axis Pointer to the angle_axis array (3x1) 
+ * @param[in] jacobian Pointer to the jacobian matrix (3x4, optional)
+ */
+static inline void quaternionToAngleAxis(const double * q, double * angle_axis, double * jacobian = nullptr) 
+{
+  ceres::QuaternionToAngleAxis(q, angle_axis);
+  if (jacobian) {
+    Eigen::Map<Eigen::Matrix<double, 3, 4, Eigen::RowMajor>> jacobian_map(jacobian);
+    const double & q0 = q[0];
+    const double & q1 = q[1];
+    const double & q2 = q[2];
+    const double & q3 = q[3];
+    const double q_sum2 = q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3;
+    const double sin_theta2 = q1 * q1 + q2 * q2 + q3 * q3;
+    const double sin_theta = std::sqrt(sin_theta2);
+    const double cos_theta = q0;
+
+    if (std::fpclassify(sin_theta) != FP_ZERO) {
```

Comment:
I followed the approach that I found for the implementation of this function in ceres:
 https://github.com/ceres-solver/ceres-solver/blob/62c03d6ff3b1735d4b0a88006486cfde39f10063/include/ceres/rotation.h#L355

If `sin_theta` is FP_SUBNORMAL then it's because also `sin_theta2` it's of the same form, meaning that some quaternion components have strange values . 
The FP_ZERO case will of course cause trouble since it's at denominator, and this is handled computing a first order taylor approximation.
The other cases will also cause problems in the outputs, but then either the conversion and the jacobians doesn't make sense anymore and there should be some other kind of error handling.

Reply:
should be solved now, I add a test similar to the one used in ceres to check that the jacobians are equal even for very small rotations. For reference see: https://github.com/ceres-solver/ceres-solver/blob/c29b5257e23f91d6a47c4db9d57350ed4985ea46/internal/ceres/rotation_test.cc#L287-L296